File: clients/client-py/queen/streams/runtime/cycle.py
"""
Build and submit a /streams/v1/cycle commit. 1:1 port of cycle.js.
"""

from __future__ import annotations

from typing import Any, Optional


class StreamsCycleFailed(Exception):
    code = "STREAMS_CYCLE_FAILED"

    def __init__(self, message: str, body: Any = None):
        super().__init__(message)
        self.body = body


async def commit_cycle(
    http,
    query_id: str,
    partition_id: str,
    consumer_group: str,
    state_ops: Optional[list] = None,
    push_items: Optional[list] = None,
    ack: Any = None,
    release_lease: bool = True,
) -> dict:
    body = {
        "query_id": query_id,
        "partition_id": partition_id,
        "consumer_group": consumer_group,
        "state_ops": state_ops or [],
        "push_items": push_items or [],
        "ack": ack,
        # Default true preserves the historical "atomic full-batch" cycle.
        # Pass False from gate operators that want to keep the source lease
        # alive so the un-acked tail of the batch is redelivered in order.
        "release_lease": False if release_lease is False else True,
    }
    res = await http.post("/streams/v1/cycle", body)
    if not res or res.get("success") is False:
        msg = res.get("error") if isinstance(res, dict) and res.get("error") else "cycle commit failed"
        raise StreamsCycleFailed(msg, body=res)
    return res

File: clients/client-js/client-v2/streams/runtime/cycle.js
/**
 * Build and submit a /streams/v1/cycle commit.
 *
 * Inputs: the full cycle context for one (query_id, source partition_id):
 *   - state_ops: upserts/deletes computed by the reducer
 *   - push_items: sink emissions (queue/partition/payload triples)
 *   - ack: source ack (transactionId, leaseId, status, optional error)
 *
 * The broker commits the cycle as one command, all or nothing; on success the
 * SDK considers the source messages safely processed. On failure, nothing of
 * the cycle is applied and the source messages remain visible for redelivery
 * via Queen's existing lease/retry.
 */

export async function commitCycle({ http, queryId, partitionId, consumerGroup, stateOps, pushItems, ack, releaseLease }) {
  const body = {
    query_id: queryId,
    partition_id: partitionId,
    consumer_group: consumerGroup,
    state_ops: stateOps || [],
    push_items: pushItems || [],
    ack: ack || null,
    // Default true preserves the historical "atomic full-batch" cycle.
    // Pass false from gate operators that want to keep the source lease
    // alive so the un-acked tail of the batch is redelivered in order.
    release_lease: releaseLease === false ? false : true
  }
  const res = await http.post('/streams/v1/cycle', body)
  if (!res || res.success === false) {
    const e = new Error(res && res.error ? res.error : 'cycle commit failed')
    e.code = 'STREAMS_CYCLE_FAILED'
    e.body = res
    throw e
  }
  return res
}
